[6.x] Render clickable hyperlinks in console

sebdesign opened the pull request:
![console-hyperlinks-1x](https://user-images.githubusercontent.com/667144/70480144-9e127a00-1ae7-11ea-8219-47c91ba0f3c5.png)

Since Symfony 4.3 supports [console hyperlinks](https://symfony.com/blog/new-in-symfony-4-3-console-hyperlinks), we can use them on the commands in the framework and in userland.

This PR implements a `hyperlink` method to render lines as hyperlinks in console commands:
```php
$this->hyperlink('https://laravel.com', 'Laravel');
```
And a helper method to format hyperlinks for inline use:
```php
$documentation = $this->getHyperlink('https://laravel.com/docs', 'documentation');
$this->info("See the {$documentation}.");
```

With that in place, we can use this feature to render hyperlinks for generated classes using commands like `make:migration`, `make:model`, etc. This is also useful for published files/directories using `vendor:publish`.

When a user click on those hyperlinks, the file will open in their editor if the terminal supports it. This feature could be enhanced by specifying the IDE and generating editor-specific links, like Ignition does, but I felt that was out of scope and did not include it in this PR.

sebdesign committed:
Render clickable hyperlinks in console

This adds a method to output a line as hyperlink to the console, and a helper method to format strings to hyperlinks, so they can be rendered inline.

sebdesign committed:
Render clickable hyperlinks for generated classes in `make` commands

When a `make` command generates a class, render the output as a hyperlink so that users can click to open the class in their editor.

sebdesign committed:
Render clickable hyperlink for the artisan:serve URL

sebdesign committed:
Render clickable hyperlinks for published files and directories

driesvints commented:
I could swear we reverted a PR which broke something with the previous implementation. We'll need to test this thoroughly before merging.

sebdesign commented:
@driesvints oh I didn't know there was a similar PR. I'll see if I can find that and get back to you.

GrahamCampbell commented:
This probably messed things up for people piping output into logs or other software,

GrahamCampbell commented:
Maybe we could only generate the hyperlinks when ansi mode is enabled and supported by the output?

sebdesign commented:
@GrahamCampbell any idea how I could find the previous implementation?

I used `hyperlink` instead of  `comment` on the `inspire` command, and added it to the scheduler with the `storeOutput` method. The log file did not have any hyperlinks.

So I assume Symfony won't render hyperlinks for non-decorated (or non-ansi?) terminals? I'm not sure how to test that in a good way.

driesvints commented:
> This probably messed things up for people piping output into logs or other software,

I think that was indeed the reason.

driesvints commented:
> I used hyperlink instead of comment on the inspire command, and added it to the scheduler with the storeOutput method. The log file did not have any hyperlinks.

Can we write an integration test for this?

sebdesign commented:
Totally, I can isolate the parts that have to do with piping and non-ansi mode and write a test for that tonight.

GrahamCampbell commented:
> I'm not sure how to test that in a good way.

Just pass`--no-ansi` when you run a command.

sebdesign committed:
Test console hyperlinks with ANSI and non-ANSI output

This test asserts that console hyperlinks are rendered correctly with ANSI output enabled and disabled.

For terminals that support ANSI output and can handle hyperlinks, the `Symfony\Component\Console\Formatter\OutputFormatterStyle` will render the hyperlinks as expected. For example, this will be the case when running `php artisan foo:hyperlink` in the terminal.

When the output of the artisan command is appended elsewhere, e.g. `php artisan foo:hyperlink >> foo.log`, Symfony will know that this output is not *decorated*, and will simply render the link text. This will also be the case when storing output from the command scheduler: `$schedule->command('foo:hyperlink')->storeOutput()`.

When PHPUnit is running, the artisan binary is *not* actually executed, but instead the console Kernel calls the command using a `BufferedOutput` instance (mocked or not). In that case the output is not decorated, so it will render the plain text and not the hyperlink. This applies for the other output commands as well, like `$this->info()`.

In order to simulate a decorated output, we will pass the `--ansi` flag to the `HyperinkCommandStub` and mark the `BufferedOutput` as decorated. This will force the `OutputFormatterStyle` to actually render the hyperlink, so we can perform an assertion on this.

sebdesign commented:
I've added a test for the ANSI behavior. Please see the detailed explanation in the commit message. While it's not an end-to-end test (there's no artisan binary in the framework, or any test that actually executes commands in the terminal), I am quite confident that this will work as expected.

Let me know if how I can improve this somehow. Cheers!

GrahamCampbell commented on tests/Integration/Console/ConsoleApplicationTest.php:
```diff
@@ -86,3 +100,19 @@ public function handle(Schedule $schedule)
+
+    public function handle()
+    {
+        if ($this->option('ansi')) {
```
This should not happen within the command.

sebdesign replied:
The reason I'm doing this in the command is because the output is not instantiated until the command is ran here:
https://github.com/laravel/framework/blob/6.x/src/Illuminate/Foundation/Testing/PendingCommand.php#L133

An option would be to override the `initialize` method in the stub:
https://github.com/symfony/console/blob/4.3/Command/Command.php#L185

Otherwise I would need to mock the output myself inside the test and run it with `$this->withoutMockingConsoleOutput()`. And then I would call `Illuminate\Console\Application::call('foo:hyperlink')` and then `Illuminate\Console\Application::output()` to perform assertions on this.

sebdesign committed:
Simplify console hyperlink tests

sebdesign commented:
@GrahamCampbell I rewrote the test without mocking or touching the `isDecorated` flag.

taylorotwell commented:
@GrahamCampbell do you have anymore concerns about this PR?

GrahamCampbell commented:
Just give me two mins, and I'll test it works how I expect on my machine. :)

GrahamCampbell commented:
Doesn't seem to actually produce a hyperlink in the terminal for `serve`?

<img width="694" alt="image" src="https://user-images.githubusercontent.com/2829600/70997473-16230600-20cd-11ea-9c79-c6e79c4ece6f.png">

taylorotwell commented:
I'm just going to hold off on this for now. I don't think it's super high priority.

sebdesign commented:
@GrahamCampbell what terminal are you using?
